src/Stratis.Bitcoin/Base/ChainHeadersBehavior.cs

```diff
@@ -131,154 +141,159 @@ protected override void DetachCore()
         {
             this.logger.LogTrace("()");
 
-            this.AttachedPeer.StateChanged -= this.AttachedNode_StateChanged;
+            this.AttachedPeer.MessageReceivedPriority -= this.AttachedPeer_MessageReceived;
+            this.AttachedPeer.StateChanged -= this.AttachedPeer_StateChanged;
 
             this.logger.LogTrace("(-)");
         }
 
-        private void Intercept(IncomingMessage message, Action continueInvocation)
+        private void AttachedPeer_MessageReceived(NetworkPeer peer, IncomingMessage message)
         {
-            this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(message), message.Message.Command, nameof(this.AttachedPeer), this.AttachedPeer?.RemoteSocketEndpoint);
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(peer), peer.RemoteSocketEndpoint, nameof(message), message.Message.Command);
 
-            var inv = message.Message.Payload as InvPayload;
-            if (inv != null)
+            switch (message.Message.Payload)
```

Comment:
Ahh I didn't know you could switch case on type 😱

Reply:
;)